src/pal/src/misc/dbgmsg.cpp

```diff
@@ -470,66 +470,80 @@ Notes :
 int DBG_printf(DBG_CHANNEL_ID channel, DBG_LEVEL_ID level, BOOL bHeader,
                LPCSTR function, LPCSTR file, INT line, LPCSTR format, ...)
 {
-    CHAR *buffer = (CHAR*)alloca(DBG_BUFFER_SIZE);
-    CHAR indent[MAX_NESTING+1];
-    LPSTR buffer_ptr;
-    INT output_size;
-    va_list args;
-    void *thread_id;
-    int old_errno = 0;
-
-    old_errno = errno;
+    struct SavedErrno
+    {
+        int saved;
+        SavedErrno() : saved(errno) { }
+        ~SavedErrno()
+        {
+            if ( saved != errno )
+            {
+                fprintf( stderr,"ERROR : errno changed from %d to %d inside DBG_printf\n", errno, saved );
+                errno = saved;
+            }
+        }
+    } saved_errno;
 
+    CHAR indent[MAX_NESTING+1];
     if(!DBG_get_indent(level, format, indent))
     {
+        // Note: we will drop log messages here if the indent gets too high, and we won't print
+        //       an error when this occurs.
         return 1;
     }
 
-    thread_id = (void *)THREADSilentGetCurrentThreadId();
+    void *thread_id = (void *)THREADSilentGetCurrentThreadId();
 
+    CHAR buffer[DBG_BUFFER_SIZE];
+    INT output_size;
     if(bHeader)
     {
         /* Print file instead of function name for ENTRY messages, because those
            already include the function name */
         /* also print file name for ASSERTs, to match Win32 behavior */
+        LPCSTR location;
         if( DLI_ENTRY == level || DLI_ASSERT == level || DLI_EXIT == level)
-        {
-            output_size=snprintf(buffer, DBG_BUFFER_SIZE,
-                                 "{%p" MODULE_FORMAT "} %-5s [%-7s] at %s.%d: ",
-                                 thread_id, MODULE_ID
-                                 dbg_level_names[level], dbg_channel_names[channel], file, line);
-        }
+            location = file;
         else
+            location = function;
+        output_size=snprintf(buffer, DBG_BUFFER_SIZE,
+                             "{%p" MODULE_FORMAT "} %-5s [%-7s] at %s.%d: ",
+                             thread_id, MODULE_ID
+                             dbg_level_names[level], dbg_channel_names[channel], location, line);
+        if( output_size < 0)
         {
-            output_size=snprintf(buffer, DBG_BUFFER_SIZE,
-                                 "{%p" MODULE_FORMAT "} %-5s [%-7s] at %s.%d: ",
-                                 thread_id, MODULE_ID
-                                 dbg_level_names[level], dbg_channel_names[channel], function, line);
+            fprintf(stderr, "ERROR : DBG_printf: snprintf header failed errno:%d (%s)\n", errno, strerror(errno));
+            output_size = 0; // don't return, just drop the header from the log message
         }
-
-        if(output_size + 1 > DBG_BUFFER_SIZE)
+        else if (output_size > DBG_BUFFER_SIZE)
```

Comment:
Yes the original condition was correct, however, the code is working a bit different now.  We are not checking to see if the data was truncated at this point...that is now checked below where you see the error "message truncated" error is printed.

This block is just making sure that `output_size` is not greater than `DBG_BUFFER_SIZE` because if it is then the length passed to `_vsnprintf_s` would be negative.  By setting `output_size` to be `DBG_BUFFER_SIZE` when it is larger, we will be passing `0` in for the length when we call `_vsnprintf_s`, but we will still know that the message was truncated because `output_size >= DBG_BUFFER_SIZE` will still be true.